{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f84db1e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Extracted G2G6S2 and A0ABR5CPA3\n",
      "Building BLAST databases...\n",
      "▶ makeblastdb -in protein-matching-IPR014184.fasta -dbtype prot -out /home/tommyor99/protein-matching-IPR014184 -parse_seqids\n",
      "\n",
      "\n",
      "Building a new DB, current time: 10/29/2025 14:29:18\n",
      "New DB name:   /home/tommyor99/protein-matching-IPR014184\n",
      "New DB title:  protein-matching-IPR014184.fasta\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /home/tommyor99/protein-matching-IPR014184\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 3384 sequences in 0.065769 seconds.\n",
      "▶ makeblastdb -in protein-matching-IPR014183.fasta -dbtype prot -out /home/tommyor99/protein-matching-IPR014183 -parse_seqids\n",
      "\n",
      "\n",
      "Building a new DB, current time: 10/29/2025 14:29:18\n",
      "New DB name:   /home/tommyor99/protein-matching-IPR014183\n",
      "New DB title:  protein-matching-IPR014183.fasta\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /home/tommyor99/protein-matching-IPR014183\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 14375 sequences in 0.278995 seconds.\n",
      " Retrieved 1000 top seed hits\n",
      "▶ blastdbcmd -db /home/tommyor99/protein-matching-IPR014184 -entry_batch seed_hits.txt\n",
      "▶ cd-hit -i seed_G2G6S2_top1000.fasta -o seed_G2G6S2_top1000.fasta_cdhit90.fasta -c 0.9 -n 5 -T 4 -M 0\n",
      "================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Aug 20 2021, 08:39:56\n",
      "Command: cd-hit -i seed_G2G6S2_top1000.fasta -o\n",
      "         seed_G2G6S2_top1000.fasta_cdhit90.fasta -c 0.9 -n 5 -T\n",
      "         4 -M 0\n",
      "\n",
      "Started: Wed Oct 29 14:29:19 2025\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 1000\n",
      "longest and shortest : 442 and 373\n",
      "Total letters: 405591\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 4 X 10M = 42M\n",
      "Table           : 2 X 65M = 130M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 173M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1500000\n",
      "Max number of word counting entries: 94770375\n",
      "\n",
      "# comparing sequences from          0  to       1000\n",
      ".---------- new table with      273 representatives\n",
      "\n",
      "     1000  finished        273  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 174M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.33\n",
      " Clustered seed hits at 90% identity\n",
      " Retrieved 100 top outgroup hits\n",
      "▶ blastdbcmd -db /home/tommyor99/protein-matching-IPR014183 -entry_batch out_hits.txt\n",
      " Combined FASTA written to combined_G2G6S2_A0ABR5CPA3.fasta\n",
      "Appended 3 fasta files to combined_G2G6S2_A0ABR5CPA3.fasta\n",
      "▶ mafft --auto --thread 4 combined_G2G6S2_A0ABR5CPA3.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nthread = 4\n",
      "nthreadpair = 4\n",
      "nthreadtb = 4\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "  301 / 376 (thread    2)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "  370 / 376\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP   349 / 375 (thread    3)\n",
      "Reallocating..done. *alloclen = 1911\n",
      "STEP   375 / 375 (thread    2)\n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "  300 / 376 (thread    2)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "  370 / 376\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP   342 / 375 (thread    3)\n",
      "Reallocating..done. *alloclen = 1920\n",
      "STEP   375 / 375 (thread    3)\n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "4 thread(s)\n",
      "\n",
      "distout=h\n",
      "rescale = 1\n",
      "  300 / 376 (thread    0)dndpre (aa) Version 7.525\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "4 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 4\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "rescale = 1\n",
      "\n",
      "  370 / 376\n",
      "Segment   1/ 13    1- 122\n",
      "002-0748-1 (thread    1) identical     001-0290-1 (thread    1) identical     002-0523-0 (thread    3) identical     \n",
      "Reached 2\n",
      "Segment   2/ 13  127- 141\n",
      "001-0748-1 (thread    2) identical     001-0049-1 (thread    4) identical     \n",
      "Converged.\n",
      "Segment   3/ 13  141- 148\n",
      "001-0748-1 (thread    4) identical     001-0291-0 (thread    4) identical     \n",
      "Converged.\n",
      "Segment   4/ 13  148- 156\n",
      "001-0748-1 (thread    1) identical     001-0440-0 (thread    3) identical     \n",
      "Converged.\n",
      "Segment   5/ 13  156- 161\n",
      "001-0748-0 (thread    3) identical     001-0551-1 (thread    4) identical     \n",
      "Converged.\n",
      "Segment   6/ 13  161- 182\n",
      "001-0748-1 (thread    2) identical     001-0353-1 (thread    2) identical     \n",
      "Converged.\n",
      "Segment   7/ 13  182- 202\n",
      "001-0748-1 (thread    2) identical     001-0517-1 (thread    4) identical     \n",
      "Converged.\n",
      "Segment   8/ 13  202- 210\n",
      "002-0748-1 (thread    1) identical     002-0052-0 (thread    1) identical     \n",
      "Converged.\n",
      "\n",
      "Reached 2\n",
      "Segment   9/ 13  210- 244\n",
      "001-0748-1 (thread    4) identical     001-0320-1 (thread    3) identical     \n",
      "Converged.\n",
      "Segment  10/ 13  244- 291\n",
      "001-0748-0 (thread    1) identical     001-0392-0 (thread    1) identical     \n",
      "Converged.\n",
      "Segment  11/ 13  291- 296\n",
      "001-0748-1 (thread    1) identical     001-0400-1 (thread    1) identical     \n",
      "Converged.\n",
      "Segment  12/ 13  296- 323\n",
      "001-0748-1 (thread    2) identical     001-0480-1 (thread    3) identical     \n",
      "Converged.\n",
      "Segment  13/ 13  323- 548\n",
      "002-0748-0 (thread    4) worse         001-0194-1 (thread    1) identical     001-0573-1 (thread    2) identical     002-0327-1 (thread    2) identical     002-0656-0 (thread    4) identical     \n",
      "Reached 2\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "4 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-i (Standard)\n",
      " Iterative refinement method (max. 2 iterations)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "FastTree Version 2.1.11 Double precision (No SSE3)\n",
      "Alignment: combined_G2G6S2_A0ABR5CPA3.fasta_aligned.fasta\n",
      "Amino acid distances: BLOSUM45 Joins: balanced Support: SH-like 1000\n",
      "Search: Normal +NNI +SPR (2 rounds range 10) +ML-NNI opt-each=1\n",
      "TopHits: 1.00*sqrtN close=default refresh=0.80\n",
      "ML Model: Whelan-And-Goldman, CAT approximation with 20 rate categories\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Alignment done: combined_G2G6S2_A0ABR5CPA3.fasta_aligned.fasta\n",
      "▶ fasttree -gamma -wag combined_G2G6S2_A0ABR5CPA3.fasta_aligned.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "      0.10 seconds: Joined    200 of    359\n",
      "Initial topology in 0.19 seconds\n",
      "Refining topology: 34 rounds ME-NNIs, 2 rounds ME-SPRs, 17 rounds ML-NNIs\n",
      "      0.20 seconds: ME NNI round 1 of 34, 301 of 360 splits, 69 changes (max delta 0.022)\n",
      "      0.36 seconds: SPR round   1 of   2, 101 of 722 nodes\n",
      "      0.56 seconds: SPR round   1 of   2, 301 of 722 nodes\n",
      "      0.69 seconds: SPR round   1 of   2, 501 of 722 nodes\n",
      "      0.89 seconds: SPR round   1 of   2, 701 of 722 nodes\n",
      "      1.02 seconds: SPR round   2 of   2, 101 of 722 nodes\n",
      "      1.15 seconds: SPR round   2 of   2, 301 of 722 nodes\n",
      "      1.33 seconds: SPR round   2 of   2, 501 of 722 nodes\n",
      "      1.51 seconds: SPR round   2 of   2, 701 of 722 nodes\n",
      "Total branch-length 22.766 after 1.60 sec\n",
      "      1.80 seconds: ML Lengths 101 of 360 splits\n",
      "      2.03 seconds: ML Lengths 201 of 360 splits\n",
      "      2.25 seconds: ML Lengths 301 of 360 splits\n",
      "      2.38 seconds: ML NNI round 1 of 17, 1 of 360 splits\n",
      "      3.13 seconds: ML NNI round 1 of 17, 101 of 360 splits, 13 changes (max delta 4.192)\n",
      "      3.91 seconds: ML NNI round 1 of 17, 201 of 360 splits, 38 changes (max delta 17.272)\n",
      "      4.67 seconds: ML NNI round 1 of 17, 301 of 360 splits, 60 changes (max delta 17.272)\n",
      "ML-NNI round 1: LogLk = -65860.502 NNIs 73 max delta 17.27 Time 5.19\n",
      "      5.24 seconds: Site likelihoods with rate category 1 of 20\n",
      "      5.37 seconds: Site likelihoods with rate category 3 of 20\n",
      "      5.51 seconds: Site likelihoods with rate category 5 of 20\n",
      "      5.62 seconds: Site likelihoods with rate category 7 of 20\n",
      "      5.80 seconds: Site likelihoods with rate category 9 of 20\n",
      "      5.93 seconds: Site likelihoods with rate category 11 of 20\n",
      "      6.06 seconds: Site likelihoods with rate category 13 of 20\n",
      "      6.16 seconds: Site likelihoods with rate category 15 of 20\n",
      "      6.27 seconds: Site likelihoods with rate category 17 of 20\n",
      "      6.37 seconds: Site likelihoods with rate category 19 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 1.000 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "      6.56 seconds: ML NNI round 2 of 17, 1 of 360 splits\n",
      "      7.00 seconds: ML NNI round 2 of 17, 101 of 360 splits, 11 changes (max delta 5.836)\n",
      "      7.63 seconds: ML NNI round 2 of 17, 201 of 360 splits, 22 changes (max delta 5.836)\n",
      "      8.25 seconds: ML NNI round 2 of 17, 301 of 360 splits, 40 changes (max delta 5.836)\n",
      "ML-NNI round 2: LogLk = -58293.690 NNIs 47 max delta 5.84 Time 8.65\n",
      "      8.65 seconds: ML NNI round 3 of 17, 1 of 360 splits\n",
      "      9.19 seconds: ML NNI round 3 of 17, 101 of 360 splits, 6 changes (max delta 8.435)\n",
      "      9.70 seconds: ML NNI round 3 of 17, 201 of 360 splits, 15 changes (max delta 8.435)\n",
      "ML-NNI round 3: LogLk = -58262.453 NNIs 17 max delta 8.44 Time 10.12\n",
      "     10.12 seconds: ML NNI round 4 of 17, 1 of 360 splits\n",
      "     10.53 seconds: ML NNI round 4 of 17, 101 of 360 splits, 7 changes (max delta 4.341)\n",
      "ML-NNI round 4: LogLk = -58251.997 NNIs 8 max delta 4.34 Time 10.89\n",
      "     10.89 seconds: ML NNI round 5 of 17, 1 of 360 splits\n",
      "ML-NNI round 5: LogLk = -58251.121 NNIs 0 max delta 0.00 Time 11.19\n",
      "Turning off heuristics for final round of ML NNIs (converged)\n",
      "     11.19 seconds: ML NNI round 6 of 17, 1 of 360 splits\n",
      "     12.01 seconds: ML NNI round 6 of 17, 101 of 360 splits, 3 changes (max delta 6.224)\n",
      "     12.78 seconds: ML NNI round 6 of 17, 201 of 360 splits, 7 changes (max delta 6.224)\n",
      "     13.60 seconds: ML NNI round 6 of 17, 301 of 360 splits, 9 changes (max delta 6.224)\n",
      "ML-NNI round 6: LogLk = -58234.206 NNIs 9 max delta 6.22 Time 14.09 (final)\n",
      "     14.08 seconds: ML Lengths 1 of 360 splits\n",
      "     14.30 seconds: ML Lengths 101 of 360 splits\n",
      "     14.51 seconds: ML Lengths 201 of 360 splits\n",
      "     14.72 seconds: ML Lengths 301 of 360 splits\n",
      "Optimize all lengths: LogLk = -58233.898 Time 14.85\n",
      "     15.54 seconds: ML split tests for    100 of    359 internal splits\n",
      "     16.26 seconds: ML split tests for    200 of    359 internal splits\n",
      "     16.99 seconds: ML split tests for    300 of    359 internal splits\n",
      "     17.47 seconds: Site likelihoods with rate category 1 of 20\n",
      "     17.61 seconds: Site likelihoods with rate category 3 of 20\n",
      "     17.74 seconds: Site likelihoods with rate category 5 of 20\n",
      "     17.87 seconds: Site likelihoods with rate category 7 of 20\n",
      "     18.00 seconds: Site likelihoods with rate category 9 of 20\n",
      "     18.13 seconds: Site likelihoods with rate category 11 of 20\n",
      "     18.27 seconds: Site likelihoods with rate category 13 of 20\n",
      "     18.40 seconds: Site likelihoods with rate category 15 of 20\n",
      "     18.53 seconds: Site likelihoods with rate category 17 of 20\n",
      "     18.67 seconds: Site likelihoods with rate category 19 of 20\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Phylogenetic tree built: combined_G2G6S2_A0ABR5CPA3.fasta_with_['P46154.fasta', 'Q9HTE3.fasta', 'A0A0H3KP42.fasta'].nwk\n",
      "\n",
      " All steps completed successfully!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "     18.79 seconds: Optimizing alpha round 1\n",
      "Gamma(20) LogLk = -58912.647 alpha = 0.395 rescaling lengths by 1.328\n",
      "Total time: 18.80 seconds Unique: 362/376 Bad splits: 0/359\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "\"\"\"\n",
    "Automated Phylogenetic Tree Builder\n",
    "-----------------------------------\n",
    "\n",
    "Paste your inputs in the section below, then just run:\n",
    "    python build_phylo_automation_simple.py\n",
    "\n",
    "This script will:\n",
    "  1. Extract your seed & outgroup protein sequences by accession\n",
    "  2. BLAST each within their respective family FASTAs\n",
    "  3. Cluster the seed results at 90% identity using CD-HIT\n",
    "  4. Combine both FASTAs\n",
    "  5. Align with MAFFT\n",
    "  6. Build a FastTree\n",
    "\n",
    "Requirements (must be installed and in PATH):\n",
    "    makeblastdb, blastp, blastdbcmd, cd-hit, mafft, fasttree\n",
    "\"\"\"\n",
    "\n",
    "import subprocess\n",
    "import sys\n",
    "import shutil\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "\n",
    "####################################\n",
    " #INPUTS#\n",
    "####################################\n",
    "\n",
    "# Seed (the protein of interest)\n",
    "SEED_ACCESSION = \"G2G6S2\"              # e.g. \"A0A123XYZ1\"\n",
    "SEED_FAMILY_FASTA = \"protein-matching-IPR014184.fasta\"\n",
    "\n",
    "# Outgroup (reference/other family)\n",
    "OUTGROUP_ACCESSION = \"A0ABR5CPA3\"      # e.g. \"WP_0987654321\"\n",
    "OUTGROUP_FAMILY_FASTA = \"protein-matching-IPR014183.fasta\"\n",
    "\n",
    "# Parameters\n",
    "SEED_TOP_HITS = 1000\n",
    "OUTGROUP_TOP_HITS = 100\n",
    "THREADS = 4\n",
    "CLUSTER_IDENTITY = 0.9  # 90%\n",
    "\n",
    "# Any extra fasta files to include in the final alignment/tree\n",
    "additional_fastas = [\n",
    "    \"P46154.fasta\",\n",
    "    \"Q9HTE3.fasta\",\n",
    "    \"A0A0H3KP42.fasta\"\n",
    "]\n",
    "\n",
    "####################################\n",
    "\n",
    "\n",
    "#Making sure the right stuff is installed\n",
    "def check_exec(name):\n",
    "    if shutil.which(name) is None:\n",
    "        sys.exit(f\" '{name}' not found \")\n",
    "\n",
    "# When we open a fasta file, we need te accession number and sequence. If we cant find it, we stop.\n",
    "def extract_record(fasta, acc):\n",
    "    with open(fasta) as f:\n",
    "        for rec in SeqIO.parse(f, \"fasta\"):\n",
    "            if acc in rec.id or acc in rec.description:\n",
    "                return rec\n",
    "    sys.exit(f\" {acc} not found in {fasta}\")\n",
    "\n",
    "#This writes the file in consistent fasta format (>ACCESSION...)\n",
    "def write_record(rec, filename):\n",
    "    with open(filename, \"w\") as f:\n",
    "        SeqIO.write(rec, f, \"fasta\")\n",
    "\n",
    "#This tells us what command is currently being run, opens the output file being made, runs the command, checks it worked\n",
    "def run_cmd(cmd, output=None):\n",
    "    print(\"▶\", \" \".join(cmd))\n",
    "    if output:\n",
    "        with open(output, \"w\") as f:\n",
    "            subprocess.run(cmd, check=True, stdout=f)\n",
    "    else:\n",
    "        subprocess.run(cmd, check=True)\n",
    "\n",
    "# Check tools\n",
    "for tool in [\"makeblastdb\", \"blastp\", \"blastdbcmd\", \"cd-hit\", \"mafft\", \"fasttree\"]:\n",
    "    check_exec(tool)\n",
    "\n",
    "# Extract sequences\n",
    "seed_rec = extract_record(SEED_FAMILY_FASTA, SEED_ACCESSION)\n",
    "out_rec = extract_record(OUTGROUP_FAMILY_FASTA, OUTGROUP_ACCESSION)\n",
    "\n",
    "seed_query = f\"seed_{SEED_ACCESSION}.fasta\"\n",
    "out_query = f\"out_{OUTGROUP_ACCESSION}.fasta\"\n",
    "write_record(seed_rec, seed_query)\n",
    "write_record(out_rec, out_query)\n",
    "\n",
    "print(f\"✅ Extracted {SEED_ACCESSION} and {OUTGROUP_ACCESSION}\")\n",
    "\n",
    "# Build BLAST DBs\n",
    "seed_db = str(Path(SEED_FAMILY_FASTA).resolve().with_suffix(''))  # Remove extension\n",
    "out_db = str(Path(OUTGROUP_FAMILY_FASTA).resolve().with_suffix(''))  # Remove extension\n",
    "\n",
    "# Force rebuild BLAST DBs\n",
    "print(\"Building BLAST databases...\")\n",
    "run_cmd([\"makeblastdb\", \"-in\", SEED_FAMILY_FASTA, \"-dbtype\", \"prot\", \"-out\", seed_db, \"-parse_seqids\"])\n",
    "run_cmd([\"makeblastdb\", \"-in\", OUTGROUP_FAMILY_FASTA, \"-dbtype\", \"prot\", \"-out\", out_db, \"-parse_seqids\"])\n",
    "\n",
    "# Verify database files exist\n",
    "required_extensions = ['.phr', '.pin', '.psq']\n",
    "for ext in required_extensions:\n",
    "    if not Path(seed_db + ext).exists():\n",
    "        sys.exit(f\" BLAST database file missing: {seed_db + ext}\")\n",
    "    if not Path(out_db + ext).exists():\n",
    "        sys.exit(f\" BLAST database file missing: {out_db + ext}\")\n",
    "\n",
    "# BLAST seed\n",
    "seed_hits = f\"seed_hits.txt\"\n",
    "cmd = [\n",
    "    \"blastp\", \"-query\", seed_query, \"-db\", seed_db,\n",
    "    \"-outfmt\", \"6 sacc bitscore\", \"-max_target_seqs\", str(SEED_TOP_HITS), \"-evalue\", \"1e-5\"\n",
    "]\n",
    "res = subprocess.run(cmd, capture_output=True, text=True, check=True)\n",
    "hits = [l.split()[0] for l in res.stdout.strip().splitlines()]\n",
    "Path(seed_hits).write_text(\"\\n\".join(hits))\n",
    "print(f\" Retrieved {len(hits)} top seed hits\")\n",
    "\n",
    "# Fetch seed hit sequences\n",
    "seed_fasta = f\"seed_{SEED_ACCESSION}_top{SEED_TOP_HITS}.fasta\"\n",
    "run_cmd([\"blastdbcmd\", \"-db\", seed_db, \"-entry_batch\", seed_hits], output=seed_fasta)\n",
    "\n",
    "# Cluster seed hits\n",
    "seed_clustered = f\"{seed_fasta}_cdhit90.fasta\"\n",
    "run_cmd([\n",
    "    \"cd-hit\", \"-i\", seed_fasta, \"-o\", seed_clustered,\n",
    "    \"-c\", str(CLUSTER_IDENTITY), \"-n\", \"5\", \"-T\", str(THREADS), \"-M\", \"0\"\n",
    "])\n",
    "print(f\" Clustered seed hits at {CLUSTER_IDENTITY*100:.0f}% identity\")\n",
    "\n",
    "# BLAST outgroup\n",
    "out_hits = f\"out_hits.txt\"\n",
    "cmd = [\n",
    "    \"blastp\", \"-query\", out_query, \"-db\", out_db,\n",
    "    \"-outfmt\", \"6 sacc bitscore\", \"-max_target_seqs\", str(OUTGROUP_TOP_HITS), \"-evalue\", \"1e-5\"\n",
    "]\n",
    "res = subprocess.run(cmd, capture_output=True, text=True, check=True)\n",
    "hits = [l.split()[0] for l in res.stdout.strip().splitlines()]\n",
    "Path(out_hits).write_text(\"\\n\".join(hits))\n",
    "print(f\" Retrieved {len(hits)} top outgroup hits\")\n",
    "\n",
    "# Fetch outgroup sequences\n",
    "out_fasta = f\"out_{OUTGROUP_ACCESSION}_top{OUTGROUP_TOP_HITS}.fasta\"\n",
    "run_cmd([\"blastdbcmd\", \"-db\", out_db, \"-entry_batch\", out_hits], output=out_fasta)\n",
    "\n",
    "# Combine FASTAs\n",
    "combined = f\"combined_{SEED_ACCESSION}_{OUTGROUP_ACCESSION}.fasta\"\n",
    "with open(combined, \"w\") as out:\n",
    "    for f in [seed_clustered, out_fasta]:\n",
    "        for rec in SeqIO.parse(f, \"fasta\"):\n",
    "            out.write(f\">{rec.id}\\n{str(rec.seq)}\\n\")\n",
    "print(f\" Combined FASTA written to {combined}\")\n",
    "\n",
    "# Append additional fasta files in the directory to the combined fasta file\n",
    "\n",
    "def append_fastas_to_combined(combined_fasta, fasta_files):\n",
    "    with open(combined_fasta, \"a\") as out:\n",
    "        for fasta in fasta_files:\n",
    "            for rec in SeqIO.parse(fasta, \"fasta\"):\n",
    "                out.write(f\">{rec.id}\\n{str(rec.seq)}\\n\")\n",
    "    print(f\"Appended {len(fasta_files)} fasta files to {combined_fasta}\")\n",
    "\n",
    "\n",
    "append_fastas_to_combined(combined, additional_fastas)\n",
    "\n",
    "# MAFFT alignment\n",
    "aligned = f\"{combined}_aligned.fasta\"\n",
    "run_cmd([\"mafft\", \"--auto\", \"--thread\", str(THREADS), combined], output=aligned)\n",
    "print(f\" Alignment done: {aligned}\")\n",
    "\n",
    "# FastTree\n",
    "tree_file = f\"{combined}_with_{additional_fastas}.nwk\"\n",
    "run_cmd([\"fasttree\", \"-gamma\", \"-wag\", aligned], output=tree_file)\n",
    "print(f\" Phylogenetic tree built: {tree_file}\")\n",
    "\n",
    "print(\"\\n All steps completed successfully!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "034042e1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
